core/src/engine/impl/wait_list_light.cpp

```diff
@@ -1,81 +1,82 @@
-  UASSERT(!wait_list_.owner_);
-  wait_list_.owner_ = &task;
-}
-
-WaitListLight::SingleUserGuard::~SingleUserGuard() {
-  auto& task = engine::current_task::GetCurrentTaskContext();
-  UASSERT(wait_list_.owner_ == &task);
-  wait_list_.owner_ = nullptr;
-}
-#endif
+struct WaitListLight::Impl final {
+  std::atomic<concurrent::impl::TaggedPtr<TaskContext>> waiter{nullptr};
+};
 
-WaitListLight::~WaitListLight() = default;
+WaitListLight::WaitListLight() noexcept : impl_() {}
 
-bool WaitListLight::IsEmpty() const { return waiting_; }
+WaitListLight::~WaitListLight() {
+  UASSERT_MSG(impl_->waiter.load().GetDataPtr() == nullptr,
+              "Someone is waiting on WaitListLight while it's being destroyed");
+}
 
-void WaitListLight::Append(
-    boost::intrusive_ptr<impl::TaskContext> context) noexcept {
+void WaitListLight::Append(boost::intrusive_ptr<TaskContext> context) noexcept {
   UASSERT(context);
   UASSERT(context->IsCurrent());
-  UASSERT(!waiting_);
-#ifndef NDEBUG
-  UASSERT(!owner_ || owner_ == context.get());
-#endif
-
+  UASSERT_MSG(!impl_->waiter.load(std::memory_order_relaxed).GetDataPtr(),
+              "Another coroutine is already waiting on this WaitListLight");
   LOG_TRACE() << "Appending, use_count=" << context->use_count();
-  waiting_ = context.detach();
+
+  // We can only fit 16-bit epoch into the atomic. This means that we allow no
+  // more than 2^16-1 coroutine sleep sessions during a single racy WakeupOne,
+  // as opposed to the regular 2^32-1 ones. Not a big deal, we use
+  // boost::lockfree queues with 16-bit tags anyway.
+  const auto cut_epoch = static_cast<std::uint16_t>(context->GetEpoch());
+
+  // Keep a reference logically stored in the WaitListLight to ensure that
+  // WakeupOne can complete safely in parallel with the waiting task being
+  // cancelled, Remove-d and stopped.
+  auto* const ctx = context.detach();
+
+  impl_->waiter.store({ctx, cut_epoch}, std::memory_order_seq_cst);
 }
 
 void WaitListLight::WakeupOne() {
-  ++in_wakeup_;
-  utils::FastScopeGuard guard([this]() noexcept { --in_wakeup_; });
-
-  auto* old = waiting_.exchange(nullptr);
-  if (old) {
-    LOG_TRACE() << "Waking up! use_count=" << old->use_count();
-    old->Wakeup(impl::TaskContext::WakeupSource::kWaitList,
-                impl::TaskContext::NoEpoch{});
-    intrusive_ptr_release(old);
-  }
+  const auto context_and_epoch = impl_->waiter.exchange(nullptr);
+  const boost::intrusive_ptr<TaskContext> context{
+      context_and_epoch.GetDataPtr(),
+      /*add_ref=*/false};
+  if (!context) return;
+
+  const auto cut_epoch = context_and_epoch.GetTag();
+  // We cheat a bit, stealing the higher 16 bits from the current epoch. This
+  // means that we only have 16 bits of epoch protection here, again. (See also
+  // comments in Append.)
+  const auto epoch = static_cast<SleepState::Epoch>(
+      (utils::UnderlyingValue(context->GetEpoch()) & 0xffff0000) | cut_epoch);
```

Comment:
Assert that there's no more than 3k (5% of 16bit value) increments from cut_epoch to actual current epoch. In release we may have 10 times better performance, so 5% seems to be kinda reasonable. Add a message on assert failure